src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs

```diff
@@ -1547,5 +1622,40 @@ internal override SyntaxNode ScopeDesignator
                 return _syntax;
             }
         }
+
+        private MethodArgumentInfo GetSpecialTypeMemberInfo(SpecialMember member, SyntaxNode syntax, DiagnosticBag diagnostics)
+        {
+            return (MethodSymbol)GetSpecialTypeMember(member, diagnostics, syntax) is { } resolvedMember
+                    ? MethodArgumentInfo.ParameterlessMethod(resolvedMember)
+                    : null;
+        }
+
+        /// <param name="extensionReceiverOpt">If extension methods are allowed, this must be non-null.</param>
+        private MethodArgumentInfo BindDefaultArguments(MethodSymbol method, BoundExpression extensionReceiverOpt, bool allowExtensionMethods, bool expanded, SyntaxNode syntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(!method.IsExtensionMethod || allowExtensionMethods);
+            Debug.Assert(extensionReceiverOpt != null || !allowExtensionMethods);
+
+            var argsBuilder = ArrayBuilder<BoundExpression>.GetInstance(method.ParameterCount);
```

Comment:
Can we avoid grabbing a builder when no args? 
Something like `var args = method.IsExtensionMethod ? ImmutableArray.Create(extensionReceiverOpt) : ImmutableArray.Empty;`.

Consider adding assertion on `method.ParameterCount` (zero unless extension, right?)

Reply:
> zero unless extension, right?

No, consider `GetAsyncEnumerator(CancellationToken = default)`. We could check the parameter count and bail early for the 0 case though.

---
In reply to: [549493609](https://github.com/dotnet/roslyn/pull/50105#discussion_r549493609) [](ancestors = 549493609)